### ISSUE TITLE
White Card 27240:501(c)(3): Investigate and resolve issue to allow unapproved 501(c)(3) to view Giving Widget correctly in Preview page

### ISSUE THREAD

**Issue by dsgallagher** [dsgallagher/Testing-GitHub-Issues#2]
Overview:
This task is to investigate and resolve a number of issues that are occurring when an unapproved 501(c)(3) nonprofit accesses its Giving Widget configuration page and tries to view the configuration in the Preview page. Essentially, an unapproved 501(c)(3) nonprofit will see an empty Giving Widget on the Preview page when their organization is the sole organization selected by default.

Business reasoning: This issue was orginally reported as a bounceback as part of linked card "White Card 27161:Nonprofit: Allow nonprofits to make guest donations through the Giving Widget":

Rafal: 11/21/2017:
have a claimed 501c3, confirm email (not approved)
go to the giving widget page and click on View 
Result: widget aint populated with the claimed 501c3 ( the direct url https://www.qanightly1.site/nonprofit/1311/givingwidget works tho ) 
also when I login as a acharity1@ammadoqa.com / password and click on View widget - it works fine
Expected: It should display on preview right away 

------------------------------------------------------------------------------

Ivan (Serbia) responded that the issue would require investigation and a new card:

Ivan Mihajlovic 5 December 2017
Issue reported by Rafal isn't something that can be changed / fixed easily. As a matter of fact this doesn't work for GW1 and since this use case wasn't defined in the UAT's I disagree that this is expected behavior. I tried changing some things in code but then was faced with a lot of bugs. I found a a few bugs nightly1 on this page when you try to add unclaimed 501c3 and then save the configuration. At the end, I spent to much time on this issue, I think that we need to investigate how much time it will take to implement this functionality and probably should be done as a new white card.

------------------------------------------------------------------------------

Damien reviewed the case and opinion and agreed to create this card with the aim of scope the efforts to fix the issues and get them resolved.

DG 05 Dec 2017
Opinion from Ivan is that this fix requires investigation and a new card. This is not a critical issue so we can proceed with creating a new card for a future iteration and getting the issue evalauted and scored at that time. 

------------------------------------------------------------------------------

**UATs**

Dependencies:

- White Card 27161:Nonprofit: Allow nonprofits to make guest donations through the Giving Widget

Preconditions: 

- Nonprofit is a 501(c)(3) that has claimed the profile but has NOT yet been approved in Admin site 

Deliverables:

- As per card "27161", any logged in nonprofit should be able to access its Giving Widget Configuration page and click the "View Widget" button to view its correctly configured Giving Widget on the Preview page - this should also happen for an unapproved 501(c)(3) nonprofit when their organization is the sole organization selected by default.

UATs:

1. Access the profile page of an "unclaimed" 501 nonprofit

- Example: ANCIENT ORDER OF CROAKING FROGS (URL: https://www.ammadonightly4.com/nonprofits/501c3/232593392)

2. Click the Manage Profile button option > complete the form

- It does not matter whether you click the link in the email confirmation or not - unconfirmed and confirmed cases should both be able to access the Giving Widget Configuration page and click the "View Widget" button to view the configured Giving Widget correctly on the Preview page when their organization is the sole organization selected by default.

3. Access menu "donations > giving widget" > scroll to "Customize your Giving Widget" section:

- The tab in the selection component "Selected (1)", which is correct, as for enlisted nonprofits it "should" prepopulate with the nonprofit itself 

- This is a logical and safe assumption that the nonprofit wants to receive donations to itself from the widget that it owns

4. The first issue noticed is when the user clicks the View Widget button to access the preview page

- Because the widget configuration page shows the selection component "automatically" populated with the 501 nonprofit, the Preview page should therefore display the widget populated with the 501(c)(3) nonprofit but it does not
- NOTE: As per conversation in Skype, for unapproved 501 we need to display the "staging nonprofit" on the preview page
- Ivan M: Adding 501c3 from whitelist component on widget configuration page, as default beneficiary, was never implemented in code. I found some :TODO comments about that. creategivingwidget.js - Line number 153 (//@TODO: Display staging nonprofit on preview page) If you select unclaimed 501c3 it will not be posted to preview page
- Only after the nonprofit application to claim its profile is approved in the Admin section will it display correctly
- I can also get the Preview page to work correctly by doing the following:
-- Remove the currently selected beneficiary (ANCIENT ORDER OF CROAKING FROGS) in the selection component
-- Search for "ANCIENT ORDER OF CROAKING FROGS"
-- Select this
-- Proceed to the Preview page and see this correctly
-- From looking at the URLs, I am seeing that the entity URL for "my nonprofit" logged in is https://www.ammadonightly4.com/nonprofit/us-232593392 BUT the entity displaying in the search results (which is also "my nonprofit") is actually slightly different as https://www.ammadonightly4.com/nonprofits/501c3/232593392 so seems search does not know this is still "my nonprofit" and is treating the search result nonprofit differently to the default selected nonprofit.
- Note - selected colour and size configuration options are displaying correctly in the Preview page
- Note - widget is prepopulating correctly when accessing the "/givingwidget" link e.g. https://www.ammadonightly4.com/nonprofit/us-232593392/givingwidget

5. The second issue noticed is when I click the "Search beneficiaries" tab in the selection component on the widget configuration page and then run a search for my own nonprofit (ANCIENT ORDER OF CROAKING FROGS)

- The search results display ANCIENT ORDER OF CROAKING FROGS as "unselected" - this is wrong; it should display in the results as "selected"
- I can select "ANCIENT ORDER OF CROAKING FROGS" and add it to the "selected" beneficiaries - as per above, I am seeing that the entity URL for "my nonprofit" logged in is https://www.ammadonightly4.com/nonprofit/us-232593392 BUT the entity displaying in the search results (which is also "my nonprofit") is actually slightly different as https://www.ammadonightly4.com/nonprofits/501c3/232593392 so seems search does not know this is also "my nonprofit".
- The result of selecting "ANCIENT ORDER OF CROAKING FROGS" from the search results and adding it to the "selected" beneficiaries is that there are now two of the same nonprofit in the "selected" beneficiaries - obviously this is wrong; you cannot add the same nonprofit twice PLUS, if I save this configuration, I will receive a 500 server error as follows: 

System.ServiceModel.FaultException`1[Ammado.ContractsV2.Common.DataTypes.CommandParamValidationFault]: Request.RestrictionListId Must be greater than zero (Fault Detail is equal to Ammado.ContractsV2.Common.DataTypes.CommandParamValidationFault).

Use case ends

Data Populator Requirements: None
Feature Management Requirements: None
Privacy Requirements: None
Activity Feeds/Email Notifications Requirements: None

**Comment by dsgallagher** [comment]
DG 08 Dec 2017

To Slobodan for score and review

**Comment by rafalf** [comment]
@dsgallagher Looks nice and clean :) says creepy bunny 
![bunny](https://user-images.githubusercontent.com/445435/34256064-68b9cf02-e64b-11e7-96ef-2502673f64cb.jpg)
